// @kintone/rest-api-client/esm/client/BulkRequestClient.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { buildPath } from "../url";
var BulkRequestClient = /** @class */ (function () {
    function BulkRequestClient(client, guestSpaceId) {
        this.client = client;
        this.guestSpaceId = guestSpaceId;
    }
    BulkRequestClient.prototype.send = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "bulkRequest"
        });
        return this.client.post(path, params);
    };
    BulkRequestClient.prototype.buildPathWithGuestSpaceId = function (params) {
        return buildPath(__assign(__assign({}, params), { guestSpaceId: this.guestSpaceId }));
    };
    return BulkRequestClient;
}());
export { BulkRequestClient };
