// @kintone/rest-api-client/esm/client/RecordClient.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
import { buildPath } from "./../url";
var RecordClient = /** @class */ (function () {
    function RecordClient(client, guestSpaceId) {
        this.client = client;
        this.guestSpaceId = guestSpaceId;
        this.didWarnMaximumOffsetValue = false;
    }
    RecordClient.prototype.getRecord = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "record"
        });
        return this.client.get(path, params);
    };
    RecordClient.prototype.addRecord = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "record"
        });
        return this.client.post(path, params);
    };
    RecordClient.prototype.updateRecord = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "record"
        });
        return this.client.put(path, params);
    };
    // TODO: `records` type in return type should be filtered by `fields`.
    RecordClient.prototype.getRecords = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "records"
        });
        if (params.query) {
            var regexp = /offset\s+(\d+)/i;
            var result = params.query.match(regexp);
            if (!this.didWarnMaximumOffsetValue &&
                result &&
                Number(result[1]) > 10000) {
                this.didWarnMaximumOffsetValue = true;
                console.warn("Warning: The maximum offset value will be limited to 10,000 in the future. Please use `createCursor()` and `getRecordsByCursor()` instead.");
            }
        }
        return this.client.get(path, params);
    };
    RecordClient.prototype.addRecords = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "records"
        });
        return this.client.post(path, params);
    };
    RecordClient.prototype.updateRecords = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "records"
        });
        return this.client.put(path, params);
    };
    RecordClient.prototype.deleteRecords = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "records"
        });
        return this.client.delete(path, params);
    };
    RecordClient.prototype.createCursor = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "records/cursor"
        });
        return this.client.post(path, params);
    };
    RecordClient.prototype.getRecordsByCursor = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "records/cursor"
        });
        return this.client.get(path, params);
    };
    RecordClient.prototype.deleteCursor = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "records/cursor"
        });
        return this.client.delete(path, params);
    };
    RecordClient.prototype.getAllRecords = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var condition, orderBy, _a, withCursor, rest, conditionQuery, query;
            return __generator(this, function (_b) {
                condition = params.condition, orderBy = params.orderBy, _a = params.withCursor, withCursor = _a === void 0 ? true : _a, rest = __rest(params, ["condition", "orderBy", "withCursor"]);
                if (!orderBy) {
                    return [2 /*return*/, this.getAllRecordsWithId(__assign(__assign({}, rest), { condition: condition }))];
                }
                if (withCursor) {
                    conditionQuery = condition ? condition + " " : "";
                    query = "" + conditionQuery + (orderBy ? "order by " + orderBy : "");
                    return [2 /*return*/, this.getAllRecordsWithCursor(__assign(__assign({}, rest), { query: query }))];
                }
                return [2 /*return*/, this.getAllRecordsWithOffset(__assign(__assign({}, rest), { orderBy: orderBy, condition: condition }))];
            });
        });
    };
    RecordClient.prototype.getAllRecordsWithId = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var originalFields, rest, fields;
            return __generator(this, function (_a) {
                originalFields = params.fields, rest = __rest(params, ["fields"]);
                fields = originalFields;
                // Append $id if $id doesn't exist in fields
                if (fields && fields.length > 0 && fields.indexOf("$id") === -1) {
                    fields = __spreadArrays(fields, ["$id"]);
                }
                return [2 /*return*/, this.getAllRecordsRecursiveWithId(__assign(__assign({}, rest), { fields: fields }), 0, [])];
            });
        });
    };
    RecordClient.prototype.getAllRecordsRecursiveWithId = function (params, id, records) {
        return __awaiter(this, void 0, void 0, function () {
            var GET_RECORDS_LIMIT, condition, rest, conditionQuery, query, result, allRecords, lastId;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        GET_RECORDS_LIMIT = 500;
                        condition = params.condition, rest = __rest(params, ["condition"]);
                        conditionQuery = condition ? condition + " and " : "";
                        query = conditionQuery + "$id > " + id + " order by $id asc limit " + GET_RECORDS_LIMIT;
                        return [4 /*yield*/, this.getRecords(__assign(__assign({}, rest), { query: query }))];
                    case 1:
                        result = _a.sent();
                        allRecords = records.concat(result.records);
                        if (result.records.length < GET_RECORDS_LIMIT) {
                            return [2 /*return*/, allRecords];
                        }
                        lastId = result.records[result.records.length - 1].$id.value;
                        return [2 /*return*/, this.getAllRecordsRecursiveWithId(params, lastId, allRecords)];
                }
            });
        });
    };
    RecordClient.prototype.getAllRecordsWithOffset = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.getAllRecordsRecursiveWithOffset(params, 0, [])];
            });
        });
    };
    RecordClient.prototype.getAllRecordsRecursiveWithOffset = function (params, offset, records) {
        return __awaiter(this, void 0, void 0, function () {
            var GET_RECORDS_LIMIT, condition, orderBy, rest, conditionQuery, query, result, allRecords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        GET_RECORDS_LIMIT = 500;
                        condition = params.condition, orderBy = params.orderBy, rest = __rest(params, ["condition", "orderBy"]);
                        conditionQuery = condition ? condition + " " : "";
                        query = "" + conditionQuery + (orderBy ? "order by " + orderBy + " " : "") + "limit " + GET_RECORDS_LIMIT + " offset " + offset;
                        return [4 /*yield*/, this.getRecords(__assign(__assign({}, rest), { query: query }))];
                    case 1:
                        result = _a.sent();
                        allRecords = records.concat(result.records);
                        if (result.records.length < GET_RECORDS_LIMIT) {
                            return [2 /*return*/, allRecords];
                        }
                        return [2 /*return*/, this.getAllRecordsRecursiveWithOffset(params, offset + GET_RECORDS_LIMIT, allRecords)];
                }
            });
        });
    };
    RecordClient.prototype.getAllRecordsWithCursor = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var id, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.createCursor(params)];
                    case 1:
                        id = (_a.sent()).id;
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 4, , 5]);
                        return [4 /*yield*/, this.getAllRecordsRecursiveByCursor(id, [])];
                    case 3: return [2 /*return*/, _a.sent()];
                    case 4:
                        error_1 = _a.sent();
                        this.deleteCursor({ id: id });
                        throw error_1;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    RecordClient.prototype.getAllRecordsRecursiveByCursor = function (id, records) {
        return __awaiter(this, void 0, void 0, function () {
            var result, allRecords;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getRecordsByCursor({ id: id })];
                    case 1:
                        result = _a.sent();
                        allRecords = records.concat(result.records);
                        if (result.next) {
                            return [2 /*return*/, this.getAllRecordsRecursiveByCursor(id, allRecords)];
                        }
                        return [2 /*return*/, allRecords];
                }
            });
        });
    };
    RecordClient.prototype.addRecordComment = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "record/comment"
        });
        return this.client.post(path, params);
    };
    RecordClient.prototype.deleteRecordComment = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "record/comment"
        });
        return this.client.delete(path, params);
    };
    RecordClient.prototype.getRecordComments = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "record/comments"
        });
        return this.client.get(path, params);
    };
    RecordClient.prototype.updateRecordAssignees = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "record/assignees"
        });
        return this.client.put(path, params);
    };
    RecordClient.prototype.updateRecordStatus = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "record/status"
        });
        return this.client.put(path, params);
    };
    RecordClient.prototype.updateRecordsStatus = function (params) {
        var path = this.buildPathWithGuestSpaceId({
            endpointName: "records/status"
        });
        return this.client.put(path, params);
    };
    RecordClient.prototype.buildPathWithGuestSpaceId = function (params) {
        return buildPath(__assign(__assign({}, params), { guestSpaceId: this.guestSpaceId }));
    };
    return RecordClient;
}());
export { RecordClient };
